add Stdlib.Float.Array

damiendoligez opened the pull request:
Work in progress:
* ~need to add tests~
* ~need to discuss introducing a `Float` -> `List` dependency~
* ~maybe some more functions should be added~

damiendoligez committed:
Add Stdlib.Float.Array module with a bunch of functions for the
floatarray type.

xclerc commented:
For the sake of consistency, I would suggest to consider
the addition of a labelled version.

damiendoligez committed:
add labelled version

damiendoligez committed:
add tests and fix bug in concat

damiendoligez commented:
I've added the labeled version, I've added tests (which revealed a bug in `concat`) and I've decided that more functions can be added in a later PR.

This is now ready for review.

Two points that may need discussion:
* the dependency on `List` introduced by this PR
* the reuse of `memq` with a different (but related) meaning

@xclerc, @alainfrisch would you like to review?

alainfrisch commented on stdlib/float.mli:
```diff
@@ -257,11 +257,264 @@ val hash: t -> int
+
+  val memq : float -> t -> bool
+  (** Same as {!mem}, but uses float equality instead of generic
+      equality to compare elements. *)
```
The doc for mem and memq is confusing (or it is me being confused, perhaps).  "float equality" could be interpreted as Float.equal, which behaves as the generic comparison (i.e. Stdlib.compare x y = 0), but memq rather uses the generic equality (i.e. (=)).

mshinwell replied:
I'm not sure we're in a good state at the moment with regard to IEEE semantics on float equality.  I tend to think that it's a mistake that `Float.equal` does not have the IEEE semantics---my guess is that users would expect it to have such semantics (I certainly did), and indeed the polymorphic equality operator has such semantics.  I wonder if we should change `Float.equal` and add a second `Float.equal_non_ieee` or similar.  I know this means that `equal` functions cannot in general be defined as `compare x y = 0`, but I think that is an inevitable consequence if assuming IEEE semantics globally.

yallop replied:
It looks as though `Array.(memq x (make 1 x))` is always false for floats, but `Float.Array.(memq x (make 1 x))` is true (for non-NaNs).

That seems rather surprising, and I wonder if it'd be better to pick a different name.  Alternatively, perhaps just drop `memq` altogether, since `memq x` is just `exists ((=)x)` in any case.

alainfrisch commented:
I'll review that in more details later, but one high-level suggestion would be to use this PR to start thinking about a generic `ARRAY` signature such that `Float.Array` implements `ARRAY with type elt = float and type t = floatarray`.    Typically, one will want to have `Bytes` implement `ARRAY with type elt = char and type t = bytes`, which could create some constraints in terms of naming values.

(Also, it could make sense to make sure that one could map between two specialized array types, without going through generic arrays (perhaps by passing the implementations as first class modules).)

mshinwell commented:
@alainfrisch I agree there should be a standard signature, although I would prefer to see "`Array_intf.S`" rather than the rather shouty retro-SML "`ARRAY`".

xclerc commented on stdlib/float.mli:
```diff
@@ -257,11 +257,264 @@ val hash: t -> int
+  val get : t -> int -> float
+  (** [get a n] returns the element number [n] of floatarray [a].
+
+      Raise [Invalid_argument "index out of bounds"]
```
The documentation in `array.mli` no longer mentions the string
argument to `Invalid_argument`.

xclerc commented on stdlib/float.mli:
```diff
@@ -257,11 +257,264 @@ val hash: t -> int
+  (** [set a n x] modifies floatarray [a] in place, replacing element
+      number [n] with [x].
+
+      Raise [Invalid_argument "index out of bounds"]
```
(Same issue.)

xclerc commented on stdlib/float.mli:
```diff
@@ -257,11 +257,264 @@ val hash: t -> int
+
+      Raise [Invalid_argument] if [n < 0] or [n > Sys.max_floatarray_length]. *)
+
+  val create : int -> t
```
It feels a bit odd to have `Array.create` deprecated
while `Float.Array.create` is not. (That said, I reckon
that marking a function just introduced as deprecated
might feel equally weird.)

damiendoligez replied:
The general idea with `make` and `create` is that `make` builds initialized data structures while `create` allocates uninitialized data.

In this light, `Array.create` should not exist while `Float.Array.create` is a valid operation.

alainfrisch replied:
(That said, allocating an uninitialized array should be somehow discouraged for casual use (perhaps just with a disclaimer in the docstring), for evident reasons (non-reproducibility, risk of leaking sensitive information, etc).    The same applies to Bytes.create of course.)

xclerc replied:
Understood.

xclerc commented on stdlib/float.mli:
```diff
@@ -257,11 +257,264 @@ val hash: t -> int
+
+      Raise [Invalid_argument] if [n < 0] or [n > Sys.max_floatarray_length]. *)
+
+  (* create_float *)
```
Not having `create_float` seems natural, but it means
that `Array` and `Float.Array` will not share the very same
signature. (Anyway, this comment should probably go.)

damiendoligez replied:
`Array` and `Float.Array` cannot have the same signature anyway because in one case `t` is a type constructor and in the other it's a simple type.

xclerc replied:
Indeed

xclerc commented on stdlib/float.mli:
```diff
@@ -257,11 +257,264 @@ val hash: t -> int
+
+     Raise [Invalid_argument] if [n < 0] or [n > Sys.max_floatarray_length]. *)
+
+  (* make_matrix *)
```
(Same issue.)

damiendoligez replied:
While we could have `create_float` (but it would be slightly weird), we can't have arrays of arrays in `Float.Array` so `make_matrix` cannot exist.

xclerc replied:
(I was more or less expecting `val make_matrix : int -> int -> float -> t array`,
but I reckon it might be more confusing than useful.)

xclerc commented on stdlib/float.mli:
```diff
@@ -257,11 +257,264 @@ val hash: t -> int
+      Raise [Invalid_argument] if the length of [l] is greater than
+      [Sys.max_floatarray_length].*)
+
+  (* CR doligez: Should we insert subtitles as in [array.mli] ? *)
```
(I would.)

xclerc commented on stdlib/array.mli:
```diff
@@ -220,8 +226,7 @@ val sort : ('a -> 'a -> int) -> 'a array -> unit
    complete specification).  For example, {!Stdlib.compare} is
-   a suitable comparison function, provided there are no floating-point
-   NaN values in the data.  After calling [Array.sort], the
+   a suitable comparison function.  After calling [Array.sort], the
```
(Why not keep this sentence? I guess developers
are likely to use `float array` values for some time,
*e.g.* to interact with legacy libraries.)

damiendoligez replied:
In fact `Stdlib.compare` is suitable even if there are NaN values.

xclerc commented on stdlib/float.mli:
```diff
@@ -257,11 +257,264 @@ val hash: t -> int
+
+      The current implementation uses Heap Sort.  It runs in constant
+      stack space.
+
```
(Don't we need to discuss NaN values?)

damiendoligez replied:
I think the section "specification of the comparison function" takes care of that by ruling out using the IEEE comparison operators.

xclerc replied:
OK

xclerc commented on stdlib/float.ml:
```diff
@@ -92,11 +92,361 @@ external seeded_hash_param : int -> int -> int -> float -> int
+    in
+    loop l 0;
+    result
+
```
Should we be worried performance-wise?
(`Array.`{`append`, `concat`} being basically
implemented in C, using `memcpy`)

damiendoligez replied:
Is memcpy that much faster? I'll do some benchmarking.

xclerc replied:
I don't know, but it might indeed be worth checking.

xclerc commented on stdlib/float.ml:
```diff
@@ -92,11 +92,361 @@ external seeded_hash_param : int -> int -> int -> float -> int
+    List.init (length a) (unsafe_get a)
+
+  let of_list l =
+    let result = create (List.length l) in
```
(It is a bit of a stretch but contrary to `concat`,
we do not protect against overflow during length
computation.)

damiendoligez replied:
Unlike the `concat` case, `List.length` cannot overflow.

alainfrisch replied:
It could with js_of_ocaml, no?

damiendoligez replied:
With 32-bit ints on a 64-bit machine, it seems so. Then we have the same problem in `array.ml`

xclerc commented on stdlib/float.ml:
```diff
@@ -92,11 +92,361 @@ external seeded_hash_param : int -> int -> int -> float -> int
+    in
+    fill 0 l
+
+  let iter f a =
```
(Duplicated, so probably deserves to be marked as so,
like functions below.)

xclerc commented on stdlib/float.ml:
```diff
@@ -92,11 +92,361 @@ external seeded_hash_param : int -> int -> int -> float -> int
+  let iter f a =
+    for i = 0 to length a - 1 do f (unsafe_get a i) done
+
+  let iter2 f a b =
```
(Same issue.)

xclerc commented on stdlib/float.ml:
```diff
@@ -92,11 +92,361 @@ external seeded_hash_param : int -> int -> int -> float -> int
+      r
+    end
+
+  let iteri f a =
```
(Same issue.)

xclerc commented on stdlib/float.ml:
```diff
@@ -92,11 +92,361 @@ external seeded_hash_param : int -> int -> int -> float -> int
+    in
+    aux 0
+
+  (* mostly duplicated from array.ml *)
```
(The "mostly" is related to the special casing of
the empty array. Shall we keep the special casing
by introducing a constant?)

xclerc replied:
(Would also apply to `map`, `map2`, and `mapi`.)

damiendoligez replied:
There's also the `create` that takes only one argument.

xclerc replied:
Sorry, I meant "mostly" because the version from `Array`
starts with: `function [] -> [||] | ...`. So, I wondered
whether it would make sens to have a similar fast track
here, possibly by defining `let empty = create 0`.

xclerc commented:
I have made a first slightly-superficial review (*e.g.* did not
read the tests yet) in order to give early feedback.

I would second the introduction of `Array_intf.S`, and if
there is no consensus at least introduce such a signature
in the testsuite in order to enforce consistency through
tests.

Also, it looks like it cannot be part of the review because
the code is untouched, but I would suggest to mark the
`Array.Floatarray` module as deprecated (even though
it is currently "undocumented").

alainfrisch commented:
> I would prefer to see "Array_intf.S" rather than the rather shouty retro-SML "ARRAY".

`Array_intf.S` also does not look very stdlib-ish.  What about `Array.S`?

xclerc commented:
(Well, module name aside, I think it makes sense to have
the signature outside of `Array` since its very purpose is
to be shared by multiple structures.)

alainfrisch commented:
> (Well, module name aside, I think it makes sense to have
the signature outside of Array since its very purpose is
to be shared by multiple structures.)

Why would defining the signature inside `Array` make it more difficult to share it by multiple structures?

Note that `Array` itself would *not* satisfy this signature (because of the type parameter).

alainfrisch commented:
(But `Array` should probably expose an `Array.Make` to produce monomorphic array structures compatible with `Array.S`.)

xclerc commented:
> Why would defining the signature inside Array make it more difficult to share it by multiple structures?

It would certainly not be more difficult; it just felt asymmetrical.

damiendoligez commented:
If we define a signature, it cannot contain `make_float` (which is a hack, so that's probably OK) and `make_matrix` (more annoying).

damiendoligez committed:
changes suggested by @xclerc

xclerc commented:
After re-reading, the whole thing, I would have
some minor clearly-non-blocking comments:

- the documentation mentions `-flat-float-array`
  (configure-time option), but I understand the name
  might change due to @shindere 's work on [autoconf](https://github.com/ocaml/ocaml/pull/2139);
- the tests could use more [nan] occurrences
  (for instance in comparison/sorting tests);
- the tests might benefit from a parallel execution
  of operations on `Float.Array.t` and `float array`
  values. The `Array` module has been "proved"
  correct by the test of time, so we could leverage
  that to test the new module (and ensure that
  implementations will not diverge at some point).

xclerc commented:
ping @damiendoligez

It looks like the only important point to discuss is https://github.com/ocaml/ocaml/pull/1936#discussion_r224159235.

damiendoligez committed:
update documentation comments for autoconf-based config

damiendoligez committed:
fix after rebase

damiendoligez committed:
clarify the documentation for `mem`

damiendoligez committed:
rename memq to mem_ieee and add some NaNs and infinities to the tests

damiendoligez commented:
I love indian food, so I added some nans. Also infinities.
I now run the same tests on both `Float.Array.t` and `float array`.
Following offline discussions with @mshinwell and @lpw25 I renamed `memq` to `mem_ieee`.
I clarified the documentation of `Array.mem` because "structural equality" might mean `=` or `compare` but they are not the same, especially with respect to `nan`.

xclerc commented:
That is great; thanks! (I always forget infinities, probably because
I find them far less confusing...).

(Regarding the last item on your [todo list](https://github.com/ocaml/ocaml/pull/1936#issue-204443723), I would say it is not really
a problem since you actually depend only on `List.`{`init`,`length`}.
So, if the dependency on `List` happens to be a problem, I guess
it will be acceptable to just copy these small functions to `float.ml`.)

damiendoligez committed:
add Changes entry

damiendoligez committed:
open marshal channel in binary mode

damiendoligez commented:
@alainfrisch @xclerc Can I get this formally approved?

damiendoligez commented:
Thanks !

gasche commented:
Merging this PR broke the post-merge CI, with the following log:

```
Running tests from 'tests/lib-floatarray' ...
 ... testing 'floatarray.ml' with 1 (native) => failed (Running program /home/ci/builds/workspace/main/flambda/true/label/ocaml-ubuntu-latest/testsuite/_ocamltest/tests/lib-floatarray/floatarray/ocamlopt.byte/floatarray.opt without any argument: command
/home/ci/builds/workspace/main/flambda/true/label/ocaml-ubuntu-latest/testsuite/_ocamltest/tests/lib-floatarray/floatarray/ocamlopt.byte/floatarray.opt 
failed with exit code 2)
 ... testing 'floatarray.ml' with 2 (bytecode) => passed
```

shindere commented:
To be more precise, it's an assertion failure at line 208 of the test
program, which says:

```
  assert (A.of_list (A.to_list a) = a);
```

gasche commented:
@damiendoligez or someone else, could you fix the test failure? We need our CI to be working properly, so I will revert this PR by the end of the week if the issue is not fixed.

xclerc commented:
Indeed, the assertion pointed by @shindere will fail
if the uninitialized array contains a `nan` value.

shindere commented:
Xavier Clerc (2018/12/05 16:58 +0000):
> Indeed, the assertion pointed by @shindere will fail
> if the uninitialized array contains a [nan] value.

Any idea how this should be fixed?

xclerc commented:
If this is actually the cause, we may just replace
any `nan` with *e.g.* `0.`.

damiendoligez commented:
Fixed (commit bbfe9039ac304431f86c8098aa4b37f59927612f) by both
* using an initialized array
* using `compare` instead of `=`

Tested on precheck ([build 166](https://ci.inria.fr/ocaml/job/precheck/166/)).

Apologies for committing this ugly test in the first place.
